# 1. Promise

- 비동기적인 작업을 수행할 때 사용하는 JS 객체
  - 비동기처리 : 특정 코드의 실행이 완료될 때까지 기다리지 않고 다음 코드를 먼저 수행하는 JS 특성을 의미

# 2. Promise의 상태

1. Pending(대기) : Promise 객체를 생성된 상태아며, 아직 작업이 수행되지 않은 상태

```javascript
new Promise((resolve, reject) => {});
```

2. Fullfilled(이행) : 작업이 성공적으로 수행된 상태, 이 때 작업 결과 값을 가지고 있음

```javascript
new Promise(function (resolve, reject) {
  resolve();
});

// 이행 상태가 되면 then()을 활용하여 결과값을 받을 수 있음
function getData() {
  return new Promise((resolve, reject) => {
    let data = 10;
    resolve(data);
  });
}

getData().then((resolvedData) => console.log(resolvedData));

//10
```

3. Rejected(실패) : 작업이 실패한 상태, 이 때 실패 이유를 가지고 있음

```javascript
new Promise(function (resolve, reject) {
  reject();
});

// resolve와 마찬가지로, 실패 상태가 되면 catch()를 활용하여 결과값을 받고 예외 처리할 수 있음

function getData() {
  return new Promise((resolve, reject) => {
    reject(new Error("This is rejected!"));
  });
}

getData().catch((err) => console.log(err));
```

# 4. try... catch... finally문

- 에러 처리를 구현하는 방법
- try 코드 블록에 포함된 문 중에서 에러가 발생하면 해당 에러는 catch 문의 err 변수에 전달되고 catch 코드 블록이 실행
- finally 코드 블록은 에러 발생과 상관없이 반드시 한 번 실행

# 3. 후속 처리 메서드 

## 3.1 then()

- then()은 작업이 성공적으로 완료되었을 때 호출
- 첫 번째 콜백 함수는 프로미스가 fullfilled 상태(resolve 함수가 호출 된 상태)가 되면 호출, 이 떄 콜백 함수는 프로미스의 비동기 처리 결과를 인수로 전달
- 두 번째 콜백 함수는 프로미스가 rejected 상태(reject 함수가 호출된 상태)가 되면 호출, 이 떄 콜백 함수는 프로미스의 에러를 인수로 전달받는다.

```javascript
// fullfilled
new Promise(resolve => resolve ('fulfilled'))
  .then(v => console.log(v), e => cosole.log(e)); // fullfilled

//rejected
new Promise((_, reject) => reject(new Error('rejected')))
  .then(v => console.log(v), e => cosole.eooro(e)); // Error : rejected

  .then((result) => {
    // 작업이 성공적으로 완료되었을 때 처리할 작업
  })
```

## 3.2 catch()

- catch 메서드의 콜백 함수는 프로미스가 rejected 상태인 경우만 호출

```jsx
//rejected
new Promise((_, reject) => reject(new Error('rejected')))
  .catch(v => console.log(v)); // Error : rejected
```

## 3.3 finally

- 프로미스의 성공 또는 실패와 상관 없이 무조건 한 번 호출


# 4. 프로미스의 에러 처리 

```jsx
const URL = "nave.com"

// 주소가 잘못되서 에러 발생
promiseGet(URL)
  .then(res => console.log(res));
  .catch(err => console.log(err)); // Error : 404
```

# 5. fetch

- 사용법이 간단하고 프로미스를 지원하기 때문에 비동기 처리를 위한 콜백 패턴의 단점에서 자유롭다.
- Ajax 통신을 하는 함수
- Promise 기반으로 만들어져 axios와 마찬가지로 데이터 다루기가 쉽고, 내장 라이브러리라는 장점

## 5.1 장점

- 자바스크립트의 내장 라이브러리 이므로 별도로 import 할 필요가 없음
- Promise 기반으로 만들어졌기 때문에 데이터 다루기 편리
- 내장 라이브러리 이기 때문에 업데이트에 따른 에러 방지 가능


## 5.2 단점

- 네트워크 에러 발생 시 response timeout이 없이 기다려야 함
- JSON으로 변환해주는 과정이 필요
- 상대적으로 axios에 비해 기능이 부족


### 5.1.1 GET 요청

```jsx
    request.get(https://github.com)
        .then(res => {
            if(!res.ok) throw new Error(res.statusText);
            return res.json();
        })
```

### 5.1.2 POST 요청

```jsx
    request.post(https://github.com,{
        userId: 1,
        title: JS,
        completed: false
    }).then(res => {
            if(!res.ok) throw new Error(res.statusText);
            return res.json();
        })
        .then(todos => console.log(todos))
        .catch(err -> console.log(err))
// {userId:1, title:"JS", completed: false, id: 201}
```

### 5.1.3 PATCH 요청

```jsx
    request.patch(https://github.com,{
        completed: ture
    }).then(res => {
            if(!res.ok) throw new Error(res.statusText);
            return res.json();
        })
        .then(todos => console.log(todos))
        .catch(err -> console.log(err))
// {userId:1, id: 1, title:"delectus aut autem", completed: ture}
```

### 5.1.4 DELETE 요청

```jsx
    request.delete(https://github.com,{
        completed: ture
    }).then(res => {
            if(!res.ok) throw new Error(res.statusText);
            return res.json();
        })
        .then(todos => console.log(todos))
        .catch(err -> console.log(err))
// { }
```



























